core/file.go

```diff
@@ -192,6 +193,15 @@ func (file *File) Contents(ctx context.Context) ([]byte, error) {
 	return contents, nil
 }
 
+func (file *File) Digest(ctx context.Context) (string, error) {
+	content, err := file.Contents(ctx)
```

Comment:
This will require loading the whole file into memory first, which is bad if it's a large file (100s of MB or greater).

We can actually just stream the file contents though which should avoid this. The existing `file.Open` method returns a `io.ReadCloser` which we can stream right into a `sha256` hasher, e.g. https://pkg.go.dev/crypto/sha256#example-New-File

Reply:
Doesn't BuildKit already have a digest that it uses for caching purposes, that we can reuse here?

Reply:
It doesn't have anything as part of its official API that exposes it.

We *could* use it's internals directly I suppose, specifically this `Checksum` func: https://github.com/moby/buildkit/blob/78f24da143968ed6fa4237c25e766b14a96df159/cache/contenthash/checksum.go#L54

We'd need to be careful to provide the right `path` arg (`file.File`) and such, but yeah I think that'd be another option.

Overall it would probably be slightly more convoluted than what I suggested above, but also slightly more reusable for other use cases (especially if it's added as a method on `Result` in [this package](https://github.com/dagger/dagger/blob/main/engine/buildkit/ref.go)), so I don't mind either approach.

Reply:
I understand the logic, I see that I can add a method `Digest` in the package to get the digest directly from the cache keys.
However I'm confused about the cache key.

Here's a snippet of my attempt:

```go
func (r *ref) Digest(ctx context.Context) (string, error) {
	cacheResult, err := r.resultProxy.Result(ctx)
	if err != nil {
		return "", err
	}

    // This is an array of `[]ExportableCacheKey` that have a method `Digest` in it, however
    // What should I do with this array? Should I sum all of them in a hasher?? I'm confused 
	keys := cacheResult.CacheKeys()
```

As explained in the code's comment, what should I do with the array? Should I only keep the last entry of the array or should I compute a sum of all digest?

The `ref.Digest` method can then be called from the file proxy service, like for evaluate so that super cool but I'm stuck on the part above.
Let me know what I should do with it.

Reply:
It seems I found a workaround with 

```go
func (r *ref) Digest(ctx context.Context, path string) (digest.Digest, error) {
	cacheRef, err := r.CacheRef(ctx)
	if err != nil {
		return "", err
	}

	sessionGroup := bksession.NewGroup(r.c.ID())

	return contenthash.Checksum(ctx, cacheRef, path, contenthash.ChecksumOpts{}, sessionGroup)
}
```